### DISCUSSION TITLE
Assembly load context issues

### DISCUSSION THREAD

**Discussion by jmerle** [dotnet/interactive#2642] [General]
**The package and version I'm asking about**: 

```
$ dotnet interactive --version
1.0.222301+add80010a18ef6fe7b3f47ce1bfc6015aec222a8
```

### Question

For an application at work we have a class which uses `Assembly.LoadFrom(<path to dll>).ExportedTypes` to create a list of all the public types that our assemblies export. This class contains a method like `GetExportedType<T>(string typeName)` to retrieve a type from this list, to construct a new instance of it and to cast it to `T`. In this method we go over all the exported types, compare the names and check `typeof(T).IsAssignableFrom(type)` to see whether there is a valid match.

For [`IsAssignableFrom`](https://docs.microsoft.com/en-us/dotnet/api/system.type.isassignablefrom?view=net-5.0) to work properly it is required for both `typeof(T).Assembly` and `type.Assembly` to be from the same assembly load context. Currently we rely on both assemblies being in the default assembly load context (i.e. [`AssemblyLoadContext.Default`](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.loader.assemblyloadcontext.default?view=net-5.0)), but that assumption breaks when using our application in a .NET Interactive C# notebook, because notebook cells use their own assembly load context. However, if we call `Assembly.LoadFrom(<path to dll>)` for all our application's assemblies, the next time a notebook cell uses our application it magically loads our application's assemblies from the default assembly load context.

Although this fixes our issue, we're wondering what's going on here and whether we can improve our workaround. Consider the following minimal example:

Lets say I have a library called `ALCPrinter` containing one class:
```csharp
using System;
using System.Reflection;
using System.Runtime.Loader;

namespace ALCPrinter
{
    public class Printer
    {
        public static void PrintALC()
        {
            Console.WriteLine(AssemblyLoadContext.GetLoadContext(Assembly.GetExecutingAssembly()));
        }
    }
}
```

Now consider the following three example notebooks:

Example 1:
![](https://i.imgur.com/YOnNpoT.png)

Example 2:
![](https://i.imgur.com/9vGktSb.png)

Example 3 after running the cell for the first time:
![](https://i.imgur.com/UeVuyzr.png)

Example 3 after running the cell for the second time:
![](https://i.imgur.com/jLdBD9X.png)

Repository: [jmerle/alc-printer](https://github.com/jmerle/alc-printer)

Why does the assembly load context differ in these examples?

Ideally I'd like it to print that it's in the default assembly load context like in example 2, but using only a single notebook cell like in examples 1 and 3. Is that possible?

**Comment by jonsequitur** [comment]
Something to keep in mind is that each submission (i.e. each time you run a cell, whether the same cell or a different cell) results in a new assembly being generated in-memory by C# scripting. Each submission's assembly references the previous, and what I'm inferring is happening is that you'll only see these references in the `DefaultAssemblyLoadContext` if they were referenced in one of those previous submissions. I'm not clear on how `#r` and `Assembly.LoadFrom` may differ from or interact with one another.

@tmat can undoubtedly add more detail.
